File: frontend/Services/AuthService.cs
using System.Net.Http.Json;
using LendQ.Client.Models;

namespace LendQ.Client.Services;

public class AuthService : IAuthService
{
    private readonly HttpClient _http;

    public AuthService(HttpClient http)
    {
        _http = http;
    }

    public string? AccessToken { get; private set; }
    public UserSummary? CurrentUser { get; private set; }
    public event Action? OnChange;

    public async Task<AuthTokenBundle> LoginAsync(string email, string password)
    {
        var response = await _http.PostAsJsonAsync("auth/login", new { email, password });
        response.EnsureSuccessStatusCode();
        var bundle = await response.Content.ReadFromJsonAsync<AuthTokenBundle>()
            ?? throw new InvalidOperationException("Login response was null.");
        AccessToken = bundle.AccessToken;
        CurrentUser = bundle.User;
        OnChange?.Invoke();
        return bundle;
    }

    public async Task SignUpAsync(string name, string email, string password, string confirmPassword)
    {
        var response = await _http.PostAsJsonAsync("auth/signup", new
        {
            name,
            email,
            password,
            confirm_password = confirmPassword
        });
        response.EnsureSuccessStatusCode();
    }

    public async Task ForgotPasswordAsync(string email)
    {
        var response = await _http.PostAsJsonAsync("auth/forgot-password", new { email });
        response.EnsureSuccessStatusCode();
    }

    public async Task ResetPasswordAsync(string token, string password, string confirmPassword)
    {
        var response = await _http.PostAsJsonAsync("auth/reset-password", new
        {
            token,
            password,
            confirm_password = confirmPassword
        });
        response.EnsureSuccessStatusCode();
    }

    public async Task<bool> RefreshAsync()
    {
        try
        {
            var response = await _http.PostAsync("auth/refresh", null);
            if (!response.IsSuccessStatusCode)
                return false;

            var bundle = await response.Content.ReadFromJsonAsync<AuthTokenBundle>();
            if (bundle is null)
                return false;

            AccessToken = bundle.AccessToken;
            CurrentUser = bundle.User;
            OnChange?.Invoke();
            return true;
        }
        catch
        {
            return false;
        }
    }

    public async Task<UserSummary?> GetCurrentUserAsync()
    {
        if (CurrentUser is not null)
            return CurrentUser;

        var refreshed = await RefreshAsync();
        if (!refreshed)
            return null;

        return CurrentUser;
    }

    public async Task LogoutAsync()
    {
        try
        {
            await _http.PostAsync("auth/logout", null);
        }
        finally
        {
            AccessToken = null;
            CurrentUser = null;
            OnChange?.Invoke();
        }
    }

    public async Task ResendVerificationAsync(string email)
    {
        var response = await _http.PostAsJsonAsync("auth/resend-verification", new { email });
        response.EnsureSuccessStatusCode();
    }

    public async Task ConfirmEmailAsync(string token)
    {
        var response = await _http.PostAsJsonAsync("auth/confirm-email", new { token });
        response.EnsureSuccessStatusCode();
    }
}

File: frontend/src/app/core/auth/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable, tap, map, catchError, of, shareReplay } from 'rxjs';
import { Router } from '@angular/router';
import { AuthTokenBundle, UserSummary, UserRole } from '../models/user.model';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private userSubject = new BehaviorSubject<UserSummary | null>(null);
  private accessToken: string | null = null;
  private refreshInFlight$: Observable<AuthTokenBundle> | null = null;

  user$ = this.userSubject.asObservable();
  isAuthenticated$ = this.user$.pipe(map(u => !!u));

  constructor(private http: HttpClient, private router: Router) {}

  get currentUser(): UserSummary | null {
    return this.userSubject.value;
  }

  get token(): string | null {
    return this.accessToken;
  }

  hasRole(role: UserRole): boolean {
    return this.currentUser?.roles.includes(role) ?? false;
  }

  hasAnyRole(roles: UserRole[]): boolean {
    return roles.some(r => this.hasRole(r));
  }

  login(email: string, password: string): Observable<AuthTokenBundle> {
    return this.http.post<AuthTokenBundle>('/api/v1/auth/login', { email, password }).pipe(
      tap(bundle => this.setSession(bundle))
    );
  }

  signup(name: string, email: string, password: string, confirm_password: string): Observable<void> {
    return this.http.post<void>('/api/v1/auth/signup', { name, email, password, confirm_password });
  }

  tryRefresh(): Observable<boolean> {
    return this.refresh().pipe(
      map(() => true),
      catchError(() => {
        this.clearSession();
        return of(false);
      })
    );
  }

  refresh(): Observable<AuthTokenBundle> {
    if (!this.refreshInFlight$) {
      this.refreshInFlight$ = this.http.post<AuthTokenBundle>('/api/v1/auth/refresh', {}).pipe(
        tap(bundle => this.setSession(bundle)),
        shareReplay(1),
        tap({ complete: () => (this.refreshInFlight$ = null), error: () => (this.refreshInFlight$ = null) })
      );
    }
    return this.refreshInFlight$;
  }

  logout(): Observable<void> {
    return this.http.post<void>('/api/v1/auth/logout', {}).pipe(
      tap(() => {
        this.clearSession();
        this.router.navigate(['/login']);
      })
    );
  }

  forgotPassword(email: string): Observable<void> {
    return this.http.post<void>('/api/v1/auth/forgot-password', { email });
  }

  resetPassword(token: string, password: string, confirm_password: string): Observable<void> {
    return this.http.post<void>('/api/v1/auth/reset-password', { token, password, confirm_password });
  }

  resendVerification(email: string): Observable<void> {
    return this.http.post<void>('/api/v1/auth/email-verification/resend', { email });
  }

  confirmEmail(token: string): Observable<void> {
    return this.http.post<void>('/api/v1/auth/email-verification/confirm', { token });
  }

  private setSession(bundle: AuthTokenBundle): void {
    this.accessToken = bundle.access_token;
    this.userSubject.next(bundle.user);
  }

  private clearSession(): void {
    this.accessToken = null;
    this.userSubject.next(null);
  }
}
